keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -264,6 +266,23 @@ state, it will either trigger [Un-reserve](#un-reserve) plugins (on failure) or
 
 *Note: This concept used to be referred to as "assume".*
 
+### VictimsSelection
+
+The plugin provides an extension point for selecting the victim pods for preemption.  Only one preempt plugin may be enabled at a time.  
+
+The selection process are performed in cooperating with scheduler and preempt plugin as follows.
+
+1. Scheduler firstly filters _potential victim nodes_ that satisfies the following two conditions.  It contains _potential victim pods_ that are allowed to be preempted by the preemptor pod's `spec.preemptionPolicy`. For example, if the preemptor pods' `spec.preemptionPolicy` is `Never`,  no victims are selected for any nodes.  The other is the preemptor pod can be schedulable when all the potential victim pods are assumed to be preempted.
+1. For each potential victim node, scheduler filters several pods as _victim candidate pods_ from potential victim pods with the plugin's help (see `SelectVictimCandidatesOnNode`).
+1. Scheduler calls scheduler extenders, if configured, to filter victim candidate pods.
+1. Scheduler picks one node from the potential victim nodes for preemption with the plugin's help (see `PickOneNodeForPreemption`).
+
+Thus, the plugin can provide two methods to customize behavior of victim pods selection.
+
+- `SelectVictimCandidatesOnNode`: The method receives a potential victim node and its potential victim pods.  And, the method returns victim candidate pods.  Please note the plugin is responsible that the returned victims can create enough room in the node for the preemptor pod.  That means, even if it can't create enough space for the preemptor pod, the scheduling cycle will not abort. Scheduler also passes a predicate function which accept victim pods and returns they can make enough room for the preemptor pod or not.  You could use this predicate function iteratively to find minimal victim pods.
+
+- `PickOneNodeForPreemption`: The method receives a map from node to victim candidate pods (it might be filtered by extenders) and returns node which should be the victim consequently.
```

Comment:
Probably, some plugin would require to score victim candidate nodes with configured score functions (i.e. `prioritizeNodes` function in generic scheduler).  Is that ok to pass the function as an argument of the function as similar to `SelectVictimCandidatesOnNode`??

Reply:
I need to understand the use case before I can start thinking of what is a reasonable solution.
Could you give an example of what kind of pods could a plugin like this return as victims?

Reply:
> Could you give an example of what kind of pods could a plugin like this return as victims?

My use cases are batch-job(ML/DL) focused ones. As I described in the original issue (https://github.com/kubernetes/kubernetes/issues/85871), my expected use cases of `PickOneNodeForPreemption ` method are:

>  - job duration (jobs started just now might be safer to preempt)
> - how long does the current epoch will be expected to be finished? (if the epoch just started and the snapshot of the last epoch snapshot available, it might be safer to preempt)
> - how much the training converged? (if well converged, the job is safer to preempt than others).
etc.

I expect to compare victim nodes with the sum/avg of these values or etc.

And, 

> Probably, some plugin would require to score victim candidate nodes with configured score functions

About this, I think that node scores would be useful in some cases below:

- for tie-breaking with node scores after selecting with the above criteria
- GPU bin-packing: currently, we implemented GPU bin-packing score with node scoring plugin.  we want to respect normalized node scores to select the final victim node in `PickOneNodeForPreemption `.

Reply:
I think preemption "scoring" should focus in minimizing the disruption, rather than best placement. But if you need best placement with regards to GPU, I think that could be re-implemented in your extension point (or it can call the necessary function).

I was reading through https://github.com/kubernetes/kubernetes/blob/84dc7046797aad80f258b6740a98e79199c8bb4d/pkg/scheduler/core/generic_scheduler.go#L721 and I think this can be re-written as a series of plugins which accept the list of nodes and returns a subset of nodes that are "equaly good" (according to their criteria) to be nominated.

I think this could allow you to have tie-breaking from your custom selection.

Reply:
> I think preemption "scoring" should focus in minimizing the disruption, rather than best placement

I agree. I also recognize that the plugin aims to make the definition of "disruption" customizable for plugins to introduce their own definition of it.

> But if you need best placement with regards to GPU, I think that could be re-implemented in your extension point (or it can call the necessary function).

Right.  But, I still think that giving scheduler's score function to the plugin would be useful in various cases.  Because kube-scheduler can configure with multiple score plugins (those are sometime OSS or in-house), then, it could be difficult to call them directly from the plugin.  

As you commented on https://github.com/kubernetes/enhancements/pull/1398#discussion_r393695836, we should think a way how to feed the function to the plugin, though...

> I think this can be re-written as a series of plugins which accept the list of nodes and returns a subset of nodes that are "equaly good" (according to their criteria) to be nominated.

Oh, thanks, good point!  Then, you mean if the last plugin returns multiple victim nodes, scheduler finally tie break with randomly or  [the current `pickOneNodeForPreemption`](https://github.com/kubernetes/kubernetes/blob/84dc7046797aad80f258b6740a98e79199c8bb4d/pkg/scheduler/core/generic_scheduler.go#L721)), right?

Then, I think we should split the plugin into two independent plugins.

- `VictimPodFilter`:  this focuses on filtering victim pods on the specific node. It will be called multiple times if multiple nodes can help preemption (in parallel perhaps).  If multiple plugins configured, it runs them one by one for each node.
- `VictimNodeFilter`: this focuses on filter victim nodes.  It receives `nodeToVictimPods` map and returns a subset of nodes that are "equally good".  If multiple plugins configured, it runs them one by one.  If multiple nodes remain, the scheduler will tie-break with randomly or  [the current `pickOneNodeForPreemption`](https://github.com/kubernetes/kubernetes/blob/84dc7046797aad80f258b6740a98e79199c8bb4d/pkg/scheduler/core/generic_scheduler.go#L721)).

WDYT?

Reply:
Additionally to what you say about `VictimNodeFilter`, `pickOneNodeForPreemption` can be rewritten as a series of plugins. That way you could disable any tie-breakers you don't need, reorder or introduce your filter at any point. But if you don't need that, let's keep it in one plugin.

Reply:
Pick **"one node"** won't fit as I don't want to limit the victims can be only chosen from one node. So `VictimNodeFilter` won't fit either.

Reply:
> Pick "one node" won't fit as I don't want to limit the victims can be only chosen from one node

That makes sense 👍  I agree that we don't need the requirement here.